sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateEncoder.scala

```diff
@@ -271,97 +277,97 @@ class RangeKeyScanStateEncoder(
 
   // Rewrite the unsafe row by replacing fixed size fields with BIG_ENDIAN encoding
   // using byte arrays.
+  // To handle "null" values, we prepend a byte to the byte array indicating whether the value
+  // is null or not. If the value is null, we write the null byte followed by a zero byte.
+  // If the value is not null, we write the null byte followed by the value.
+  // Note that setting null for the index on the unsafeRow is not feasible as it would change
+  // the sorting order on iteration.
   private def encodePrefixKeyForRangeScan(row: UnsafeRow): UnsafeRow = {
     val writer = new UnsafeRowWriter(numOrderingCols)
     writer.resetRowWriter()
     rangeScanKeyFieldsWithIdx.foreach { case (field, idx) =>
       val value = row.get(idx, field.dataType)
-      field.dataType match {
-        // endian-ness doesn't matter for single byte objects. so just write these
-        // types directly.
-        case BooleanType => writer.write(idx, value.asInstanceOf[Boolean])
-        case ByteType => writer.write(idx, value.asInstanceOf[Byte])
-
-        // for other multi-byte types, we need to convert to big-endian
-        case ShortType =>
-          val bbuf = ByteBuffer.allocate(field.dataType.defaultSize)
-          bbuf.order(ByteOrder.BIG_ENDIAN)
-          bbuf.putShort(value.asInstanceOf[Short])
-          writer.write(idx, bbuf.array())
-
-        case IntegerType =>
-          val bbuf = ByteBuffer.allocate(field.dataType.defaultSize)
-          bbuf.order(ByteOrder.BIG_ENDIAN)
-          bbuf.putInt(value.asInstanceOf[Int])
-          writer.write(idx, bbuf.array())
-
-        case LongType =>
-          val bbuf = ByteBuffer.allocate(field.dataType.defaultSize)
-          bbuf.order(ByteOrder.BIG_ENDIAN)
-          bbuf.putLong(value.asInstanceOf[Long])
-          writer.write(idx, bbuf.array())
-
-        case FloatType =>
-          val bbuf = ByteBuffer.allocate(field.dataType.defaultSize)
-          bbuf.order(ByteOrder.BIG_ENDIAN)
-          bbuf.putFloat(value.asInstanceOf[Float])
-          writer.write(idx, bbuf.array())
-
-        case DoubleType =>
-          val bbuf = ByteBuffer.allocate(field.dataType.defaultSize)
-          bbuf.order(ByteOrder.BIG_ENDIAN)
-          bbuf.putDouble(value.asInstanceOf[Double])
-          writer.write(idx, bbuf.array())
+      val isNullCol: Byte = if (value == null) 0x01.toByte else 0x00.toByte
+      val bbuf = ByteBuffer.allocate(field.dataType.defaultSize + 1)
+      bbuf.order(ByteOrder.BIG_ENDIAN)
+      bbuf.put(isNullCol)
+      if (isNullCol == 0x01.toByte) {
```

Comment:
@anishshri-db 
Maybe I wasn't very clear. What I meant was "consistently" prepending a single byte for nullability, regardless whether they are null or not. 

If the value is null, set 0x00 to the null byte and fill 0x00 for remaining bytes.
If the value is non-null, set 0x01 to the null byte and fill actual value for remaining bytes.

In any case, `[null byte][actual bytes]` per range column.

With this, you can keep the ordering even though there is null value in the middle of column. It's just that we decide in prior whether the ordering is "null first" vs "null last".

Reply:
I'm less worried about overhead on adding a byte consistently per column, unless we will have 100s of columns.

Reply:
@neilramaswamy @sahnib Does the above proposal make sense to you as well? Just to make sure I'm not missing something.

Reply:
@HeartSaVioR - that is what I am doing right. If you check line above, I am always adding the `isNullCol` byte for each column.

only the policy is inverted
if value is null, set 0x01 to null byte and fill 0x00 for remaining bytes (currently i am writing 0x00 for the next byte - but even that's not required because byte buffer allocate will already initialize those bytes to null and regardless we won't use them)

if the value is non-null, set 0x00 to null byte and fill remaining bytes to actual values

Reply:
Ah sorry you're using same bbuf with one more byte. My bad I missed this. I agree you even don't need to add 0x00 after null bit if bytebuffer initializes all remaining bytes with 0 (or whatever consistent).

Reply:
Yup done

Here is the allocate docs - 

```
Allocates a new byte buffer.
The new buffer's position will be zero, its limit will be its capacity, its mark will be undefined, and each of its elements will be initialized to zero. It will have a backing array and its array offset will be zero.
```

So I believe should be safe